kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,782 @@
+  return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+  local hosts = parse_hosts(path)
+
+  local function insert_answer(name, qtype, address)
+    if not address then
+      return
+    end
+
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(name .. ":" .. qtype, nil, function()
+      return {
+        ttl = LONG_LASTING_TTL,
+        expire = now() + LONG_LASTING_TTL,
+        {
+          name = name,
+          type = qtype,
+          address = address,
+          class = 1,
+          ttl = LONG_LASTING_TTL,
+        },
+      }, nil, LONG_LASTING_TTL
+    end)
+  end
+
+  for name, address in pairs(hosts) do
+    name = string_lower(name)
+
+    if address.ipv4 then
+      insert_answer(name, TYPE_A, address.ipv4)
+      insert_last_type(cache, name, TYPE_A)
+    end
+
+    if address.ipv6 then
+      insert_answer(name, TYPE_AAAA, address.ipv6)
+      if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+        insert_last_type(cache, name, TYPE_AAAA)
+      end
+    end
+  end
+
+  return hosts
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  opts = opts or {}
+
+  -- parse resolv.conf
+  local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
+  if not resolv then
+    log(WARN, "Invalid resolv.conf: ", err)
+    resolv = { options = {} }
+  end
+
+  -- init the resolver options for lua-resty-dns
+  local nameservers = (opts.nameservers and #opts.nameservers > 0) and
+                      opts.nameservers or resolv.nameservers
+  if not nameservers or #nameservers == 0 then
+    log(WARN, "Invalid configuration, no nameservers specified")
+  end
+
+  local r_opts = {
+    retrans     = opts.retrans or resolv.options.attempts or 5,
+    timeout     = opts.timeout or resolv.options.timeout or 2000, -- ms
+    no_random   = opts.no_random or not resolv.options.rotate,
+    nameservers = nameservers,
+  }
+
+  -- init the mlcache
+
+  -- maximum timeout for the underlying r:query() operation to complete
+  -- socket timeout * retrans * 2 calls for send and receive + 1s extra delay
+  local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans * 2 + 1 -- s
+
+  local resty_lock_opts = {
+    timeout = lock_timeout,
+    exptimeout = lock_timeout + 1,
+  }
+
+  ipc_counter = ipc_counter + 1
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
+  local ipc = {
+    register_listeners = function(events)
+      -- The DNS client library will be required in globalpatches before Kong
+      -- initializes worker_events.
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local cwid = ngx.worker.id()
+      for _, ev in pairs(events) do
+        local handler = function(data, event, source, wid)
+          if cwid ~= wid then
```

Comment:
We should add a comment for reviewer to understand this line, like:
"the event sender does not handle event since it has done something."

Reply:
https://github.com/Kong/kong/pull/12305/commits/b87cc14ab796307cf88d07c7aed08475a9bb0497 done

Reply:
done. https://github.com/Kong/kong/pull/12305/commits/dfa302dd1a52e104c3d58e8cf2b4ab49190343c9